pennylane_lightning/core/src/simulators/lightning_gpu/algorithms/AdjointJacobianGPU.hpp

```diff
@@ -0,0 +1,446 @@
+     * calculating the overlap between two given states.
+     *
+     * @param sv1s vector of statevector <sv1| (one for each observable). Data
+     * will be conjugated.
+     * @param sv2 Statevector |sv2>
+     * @param jac Jacobian receiving the values.
+     * @param scaling_coeff Generator coefficient for given gate derivative.
+     * @param num_observables the number of observables of Jacobian to update.
+     * @param param_index Parameter index position of Jacobian to update.
+     * @param tp_size Size of trainable parameters.
+     * @param device_buffer_jac_single_param a workspace buffer on the device of
+     * size num_observables elements. Data will be ignored and overwritten.
+     * @param host_buffer_jac_single_param a workspace buffer on the host of
+     * size num_observables elements. Data will be ignored and overwritten.
+     */
+    inline void
+    updateJacobian(const std::vector<StateVectorT> &sv1s,
+                   const StateVectorT &sv2, std::span<PrecisionT> &jac,
+                   PrecisionT scaling_coeff, size_t num_observables,
+                   size_t param_index, size_t tp_size,
+                   DataBuffer<CFP_t, int> &device_buffer_jac_single_param,
+                   std::vector<CFP_t> &host_buffer_jac_single_param) {
+        host_buffer_jac_single_param.clear();
+        for (size_t obs_idx = 0; obs_idx < num_observables; obs_idx++) {
+            const StateVectorT &sv1 = sv1s[obs_idx];
+            PL_ABORT_IF_NOT(sv1.getDataBuffer().getDevTag().getDeviceID() ==
+                                sv2.getDataBuffer().getDevTag().getDeviceID(),
+                            "Data exists on different GPUs. Aborting.");
+            innerProdC_CUDA_device(
+                sv1.getData(), sv2.getData(), sv1.getLength(),
+                sv1.getDataBuffer().getDevTag().getDeviceID(),
+                sv1.getDataBuffer().getDevTag().getStreamID(),
+                sv1.getCublasCaller(),
+                device_buffer_jac_single_param.getData() + obs_idx);
+        }
+        host_buffer_jac_single_param.resize(num_observables);
+        device_buffer_jac_single_param.CopyGpuDataToHost(
+            host_buffer_jac_single_param.data(),
+            host_buffer_jac_single_param.size(), false);
+        for (size_t obs_idx = 0; obs_idx < num_observables; obs_idx++) {
+            size_t idx = param_index + obs_idx * tp_size;
+            jac[idx] =
+                -2 * scaling_coeff * host_buffer_jac_single_param[obs_idx].y;
+        }
+
+        host_buffer_jac_single_param.clear();
+    }
+
+    /**
+     * @brief OpenMP accelerated application of observables to given
+     * statevectors
+     *
+     * @param states Vector of statevector copies, one per observable.
+     * @param reference_state Reference statevector
+     * @param observables Vector of observables to apply to each statevector.
+     */
+    template <class RefStateVectorT>
+    inline void applyObservables(
```

Comment:
Do we want OpenMP here when using the GPU? How was this handled in the L-GPU repo now?

Reply:
As discussed, let's remove it since we don't need to rely on OpenMP.